pkg/syncer/upsync/upsync_controller.go

```diff
@@ -0,0 +1,437 @@
+			return
+		}
+		if unstr, ok := nsObj.(*unstructured.Unstructured); !ok {
+			utilruntime.HandleError(fmt.Errorf("downstream ns expected to be *unstructured.Unstructured got %T", nsObj))
+			return
+		} else {
+			locatorHolder = unstr
+		}
+	}
+
+	upstreamLocator, locatorExists, err := shared.LocatorFromAnnotations(locatorHolder.GetAnnotations())
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	if !locatorExists || upstreamLocator == nil {
+		logger.V(4).Info("the namespace locator doesn't exist on downstream namespace.")
+		return
+	}
+
+	if upstreamLocator.SyncTarget.UID != c.syncTargetUID || upstreamLocator.SyncTarget.ClusterName != c.syncTargetClusterName.String() {
+		return
+	}
+
+	c.enqueueUpstream(gvr, &metav1.ObjectMeta{
+		Name:      downstreamObj.GetName(),
+		Namespace: upstreamLocator.Namespace,
+		Annotations: map[string]string{
+			logicalcluster.AnnotationKey: upstreamLocator.ClusterName.String(),
+		},
+	}, logging.WithObject(logger, downstreamObj), dirtyStatus)
+}
+
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), controllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting upsync workers")
+	defer logger.Info("Stopping upsync workers")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(queueKey)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key.key).WithValues("gvr", key.gvr)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if requeue, err := c.process(ctx, key.key, key.gvr); err != nil {
+		utilruntime.HandleError(fmt.Errorf("%q controller failed to sync %q (%s), err: %w", controllerName, key.key, key.gvr.String(), err))
+		c.queue.AddRateLimited(key)
+		return true
+	} else if requeue {
+		// only requeue if we didn't error, but we still want to requeue
+		c.queue.Add(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string, gvr schema.GroupVersionResource) (bool, error) {
+	logger := klog.FromContext(ctx)
+
+	clusterName, namespace, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false, nil
+	}
+
+	dirtyStatus := false
+	if dirtyStatusObj, found := c.dirtyStatusKeys.LoadAndDelete(queueKey{
+		gvr: gvr,
+		key: key,
+	}); found {
+		dirtyStatus = dirtyStatusObj.(bool)
```

Comment:
I think on error of the reconciler you should mark the key dirty again if it had been dirty before

Reply:
Fixed in commit fffff6927c99c86702c7709881a77057331950cd